File: backend-server/app/Http/Listeners/IngestPdfDataSource.php
<?php

namespace App\Http\Listeners;

use App\Http\Enums\IngestStatusType;
use App\Http\Events\PdfDataSourceWasAdded;
use App\Models\PdfDataSource;
use Exception;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Illuminate\Contracts\Queue\ShouldQueue;

class IngestPdfDataSource implements ShouldQueue
{

    /**
     * @throws GuzzleException
     * @throws Exception
     */
    public function handle($event)
    {
        if (!$event instanceof PdfDataSourceWasAdded) {
            return;
        }

        $botId = $event->getChatbotId();
        $pdfDataSourceId = $event->getPdfDataSourceId();

        try {
            /** @var PdfDataSource $pdfDataSource */
            $pdfDataSource = PdfDataSource::where('id', $pdfDataSourceId)->firstOrFail();

            $requestBody = [
                'type' => 'pdf',
                'shared_folder' => $pdfDataSource->getFolderName(),
                'namespace' => $botId,
            ];

            // Call to ingest service endpoint
            $client = new Client();
            $response = $client->request('POST', "http://llm-server:3000/api/ingest", [
                'json' => $requestBody,
                'timeout' => 200,
            ]);

            if ($response->getStatusCode() !== 200) {
                $pdfDataSource->setStatus(IngestStatusType::FAILED);
                $pdfDataSource->save();
                return;
            }

            $pdfDataSource->setStatus(IngestStatusType::SUCCESS);
            $pdfDataSource->save();

        } catch (Exception $e) {
            $pdfDataSource->setStatus(IngestStatusType::FAILED);
            $pdfDataSource->save();
            return;
        }
    }
}

File: dj_backend_server/web/listeners/ingest_pdf_data_source.py
# listeners.py
import os
import requests
from requests.exceptions import RequestException
from web.models.pdf_data_sources import PdfDataSource
from web.enums.ingest_status_enum import IngestStatusType
from web.signals.pdf_datasource_was_added import pdf_data_source_added
from django.core.exceptions import ObjectDoesNotExist
from django.dispatch import receiver

@pdf_data_source_added.connect
def ingest_pdf_datasource(sender, **kwargs):
    bot_id = kwargs['bot_id']
    pdf_data_source_id = kwargs['data_source_id']
    delete_folder_flag = kwargs['delete_folder_flag']
    ocr_pdf_file = kwargs.get('ocr_pdf_file', False)

    try:
        pdf_data_source = PdfDataSource.objects.get(id=pdf_data_source_id)
    except ObjectDoesNotExist:
        return

    request_body = {
        'type': 'pdf',
        'shared_folder': pdf_data_source.folder_name,
        'namespace': str(bot_id),
        'delete_folder_flag': delete_folder_flag,
        'ocr_pdf_file': ocr_pdf_file,
    }

    try:
        # Call to ingest service endpoint
        url = os.getenv('APP_URL') + "/api/ingest/" 
        response = requests.post(url, json=request_body, timeout=200)

        if response.status_code != 200:
            pdf_data_source.ingest_status = IngestStatusType.FAILED.value
            pdf_data_source.save()
            return

        pdf_data_source.ingest_status = IngestStatusType.SUCCESS.value
        pdf_data_source.save()

    except RequestException as e:
        pdf_data_source.ingest_status = IngestStatusType.FAILED.value
        pdf_data_source.save()
        return

